servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/FromIterablePublisher.java

```diff
@@ -33,7 +33,15 @@
 
     private final Iterable<? extends T> iterable;
 
-    FromIterablePublisher(Iterable<? extends T> iterable) {
+    @SuppressWarnings("unchecked")
+    static <T> Publisher<T> fromIterable0(Iterable<? extends T> iterable) {
```

Comment:
The example uses a request streaming and uncovered the following scenario worth discussion:

- PublisherAsBlockingIterable.onSubscribe does a request(16)
- FromIterablePublisher.request(..) loops calling hasNext+next until demand is exhausted. Each loop iteration calls subscriber.onNext(next()).
- The same thread looping in FromIterablePublisher.request is also responsible for draining the queue in PublisherAsBlockingIterable. This means that no data will be sent downstream until FromIterablePublisher exits the loop. This maybe perceived as “dead lock” because no data is sent even if the Iterable returns an element. 

Options / Workarounds
- Improvement: FromIterablePublisher can unwrap PublisherAsBlockingIterable and directly return underlying Publisher
    - Skips blocking conversion, but may not solve the general problem (if not of type PublisherAsBlockingIterable)
    - Followup: If not able to unwrap, force subscribeOn. This breaks the dependency between Subscription/Subscriber but may not always be necessary, what level is this applied at, how do we infer/accept an executor, and are there other areas in the blocking APIs with similar issues?
- Workaround: Reduce PublisherAsBlockingIterable demand to 1
    - batch sizes of 1 means we won’t have to wait for batch size to be exhausted before delivering downstream

Reply:
If it is not possible to recognize the source as potentially blocking on `hasNext()` then perhaps demand should be reduced to 1.

There doesn't seem to be a way to insert a `subscribeOn` here without using global executor which would be unacceptable for at least some customers. 

> are there other areas in the blocking APIs with similar issues?

I am concerned that there probably are and curious that none of our existing tests uncovered them. This problem reminds me of the problems users encounter with [`PipedOutputStream`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/PipedOutputStream.html)

Reply:
moved to https://github.com/apple/servicetalk/pull/2204

Reply:
captured this comment/discussion in https://github.com/apple/servicetalk/issues/2205 so we can make progress with the example 👍